_source/_posts/2021-08-02-fix-common-problems-cors.md

```diff
@@ -0,0 +1,530 @@
+
+```html
+<form id="form2">
+    <input type="text" id="puttext" name="puttext"/>
+    <input type="button" value="Put v1" onclick="onPut('v1')"/>
+    <input type="button" value="Put v2" onclick="onPut('v2')"/>
+</form>
+```
+
+This form has a text input and the two send buttons as before that call a new JavaScript function.
+
+Next, add the JavaScript funtion to `client/control.js`:
+
+```javascript
+function onPut(version) {
+    const url = "http://localhost:8000/api/" + version + "/messages/0";
+    var headers = {}
+
+    fetch(url, {
+        method : "PUT",
+        mode: 'cors',
+        headers: headers,
+        body: new URLSearchParams(new FormData(document.getElementById("form2"))),
+    })
+    .then((response) => {
+        if (!response.ok) {
+            throw new Error(response.error)
+        }
+        return response.json();
+    })
+    .then(data => {
+        document.getElementById('messages').value = data.messages;
+    })
+    .catch(function(error) {
+        document.getElementById('messages').value = error;
+    });
+}
+```
+
+This makes a PUT request sending the form parameters in the request body. Note that the URI ends in `/0`. This means that the request is to create or change the message with the identifier `0`.
+
+Next, define a PUT handler in the `main()` function of `rest/server.go`:
+
+```go
+r.PUT("/api/:version/messages/:id", PutMessage)
+```
+
+The message identifier is extracted as a path parameter.
+
+Finally, add the request handler function to `rest/server.go`:
+
+```go
+func PutMessage(c *gin.Context) {
+	version := c.Param("version")
+	id, _ := strconv.Atoi(c.Param("id"))
+	text := c.PostForm("puttext")
+	messages[id] = text
+	if version == "v2" {
+	    c.Header("Access-Control-Allow-Origin", "http://localhost:8080")
+	}
+	c.JSON(http.StatusOK, gin.H{"messages": messages})
+}
+```
+
+This updates the message from the form data and sends the new list of messages. The function also always sets the allow origin header, as we know that it is required.
+
+Now, restart the servers and load the web page. Make sure that the developer console is open. Add some text to the text input and hit the `Send v1` button.
+
+You will see a slightly different CORS error in the console:
+
+```
+Access to fetch at 'http://localhost:8000/api/v1/messages/0' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
+```
+
+This is saying that a preflight check was made, and that it didn't set the `Access-Control-Allow-Origin` header.
+
+Now, look at the console output from the API server:
+
+```txt
+[GIN] 2020/12/01 - 11:10:09 | 404 |         447ns |             ::1 | OPTIONS  "/api/v1/messages/0"
+```
```

Comment:
You might want to delete some spaces in this output so it renders everything without scrolling.